### PR TITLE
Fix double parenting.

### PR DESCRIPTION
**Pull request by siler**
The PushChildren command was always setting the `Parent` and `PreviousParent` to the same value which meant that a break (see ~line 31 of `parent_update_system`) preventing `PreviousParent` removal from executing would go off. This meant that if an entity was re-parented, it wasn't removed from its old parent's `Children` list.

If this is resolved we are not immediately breaking out of `parent_update_system` each loop. The `debug_assert!` begins going off - we've already performed the additions (see deleted code).

This change does the following:

1. Update `PushChildren` to properly take the `Parent` and set it as the `PreviousParent`, if it exists. Otherwise, it follows the previous behavior.
2. Removes the duplicate - and more complicated - `Children` update behavior in `hierarchy_maintenance_system`.

I'm curious, also, is this the wrong direction? It seems like we might want to limit `PushChildren` to solely update the `Parent` and `PreviousParent` components, leaving management of `Children` to `hierarchy_maintenance_system`. I did try this approach, but it did not function correctly and I wasn't making much progress debugging it. At the same time, re-parenting was completely broken and `hierarchy_maintenance_system` wasn't doing much, so this seems at least like a step in the right direction if that's the goal.

### CODE REVIEW

**Comment by siler** [comment]
Also, as a note, I changed the test to use Commands as I don't expect we can provide any sort of contract to people who are fiddling with the insides of the parent/child relationships.

**Comment by cart** [comment]
I've been working on some related changes (which got sidetracked due to my ECS V2 work). The idea is to make parenting completely transactional (children can only be added via World/Command builders or newly added HierarchyQuries), which means we don't need to check for duplicates, use a "hierarchy maintenance system", or maintain a `PreviousParent` component.

https://github.com/cart/bevy/tree/hierarchy/crates/bevy_transform

I think I want to wait and see how that experiment pans out. I'm curious to hear what you think!

**Comment by cart** [comment]
(it is very much a WIP at this point)

**Comment by siler** [comment]
That looks pretty cool, looks like it would solve the issue. I'm a fan of the solution for making parenting operations transactional and better change tracking will solve the transform side of things.

This PR might be worth merging in the short term just to fix parenting in cases where people re-parent, I'm definitely going to be using it locally since it is broken otherwise. Depends on when you're planning on getting back to the transforms.

I wouldn't mind helping out there a bit, but the Query code looks kind of like magic to me and I don't really understand how we are guaranteeing that we have singular write access to the two entities in the query.

Thanks for taking a look, cart!

**Comment by MinerSebas** [comment]
As the carts hierarchy changes were ultimately sidelined and the 0.5 release is imminante, I think this PR should be updated, to include it in 0.5 .

**Comment by alice-i-cecile** [comment]
> As the carts hierarchy changes were ultimately sidelined and the 0.5 release is imminante, I think this PR should be updated, to include it in 0.5 .

I agree: this is a good stop-gap fix for the release.

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs [hunk 1]
```diff
@@ -45,31 +40,7 @@ pub fn parent_update_system(
         } else {
             commands.insert(entity, PreviousParent(parent.0));
         };
-
```
Comment (cart) [inline crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:48]: Wont removing this break manual insertion of the `Parent(Entity)` component (or changing the parent value directly)? If so, that would be a pretty major breaking change.

Comment (cart) [reply crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:48]: Yup theres no question that the whole system needs an overhaul. Currently it is possible (and easy) to mutably change Parent (either via commands or &mut Parent queries). Using that to change parents has probably been done (and regardless its _possible_ and easy to do). Breaking that behavior silently feels about as problematic as the bug being fixed in this pr. If we can do it in a way that completely prevents manual Parent mutations (both via queries and direct world access), thats worth considering.

I'm also starting to consider moving forward on merging my `hierarchy` branch, or if that ends up not being ready / too disruptive for 0.5, maybe just leaving this behavior as-is for 0.5.

 Let me know if you have any other suggestions.

Comment (alice-i-cecile) [reply crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:48]: My preference would be to leave this behavior as-is and then do a full RFC on how parent-child stuff should work immediately after 0.5 lands. It's very central to UI, and @BoxyUwU's relations are also threatening to disrupt the space and need consideration.

Comment (siler) [reply crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:48]: Cool, yeah there's several sets of changes in motion that will contribute to a better solution for this problem space. I don't have any good ideas about how to fix it in a way that works for everyone with what we have now, I think we need either Boxy's changes or something like #1655. I'm good closing this then, I'll keep running a fork until we get a proper solution.

**Comment by cart** [comment]
Closing this for now in favor of a future overhaul.

**Pull request closed without merging**